### DISCUSSION TITLE
unable to deploy gocd server/agent on kubernetes

### DISCUSSION THREAD

**Discussion by hakim-dc** [gocd/gocd#12958] [Q&A]
i'm trying to deploy gocd on kubernetes using helm chart.
everything seems well configured but i'm always getting a errors in the consol.
for server: i got classNotFoundException GoLauncher (note go.jar is there)
for agent: i got also classNotFoundException AgentBootstrapper.
i tried to use the last version of gocd images on dockerhub and i'm getting always the same issue.
am i missing something, could anyone help please.

**Comment by chadlwilson** [comment]
No, that certainly shouldn't happen, nor be possible.

Are you using the official helm-chart from https://artifacthub.io/packages/helm/gocd/gocd / https://github.com/gocd/helm-chart ?

They could happen for multiple reasons, but those errors sound like you might be overwriting or mounting custom volumes into the wrong place; or perhaps using/building custom server/agent images which are broken.

Please share your configuration (`values` overrides), commands you're trying to run to deploy, and preferably type of Kubernetes cluster - it's not really possible to know the problem from the description above.

**Reply by hakim-dc** [reply]
Hello,
so this is the kubernetes vesion
Client Version: v1.25.1
Kustomize Version: v4.5.7
Server Version: v1.22.9

Note: my company are applying Open policy agent security rules (OPA rules)

these rules generally prevent the image to write directly on the pod disk (even if it is an internal configuration that need to be generated) using this code

```
securityContext:
readOnlyRootFilesystem: true
allowPrivilegeEscalation: false

```

let's start with server first, i will provide some details (i can't copy paste conf here, company block post request on github...)
i'am mounting a lot of volume to bypass the readonly rules.
here are some details about the server conf:

```
kind: Deployment
apiVersion: apps/v1
metadata:
  name: gocd-server
  namespace: gocd
  uid: 4effe32b-b04d-401d-8aa3-75f1655bd6b6
  resourceVersion: '122843691'
  generation: 33
  creationTimestamp: '2024-07-21T18:55:16Z'
  labels:
   app: gocd
   app.kubernetes.io/managed-by: Helm
   chart: gocd-2.4.1
   component: server
   heritage: Helm
   release: gocd
  annotations:
   deployment.kubernetes.io/revision: '12'
    meta.helm.sh/release-name: gocd 
    meta.helm.sh/release-namespace: gocd
spec:
  replicas: ▸
  selector:
  matchLabels:
    app: gocd
    component: server
    release: gocd
spec: 
volumes: 
 - name: config-vol 
   configMap: 
     name: gocd 
     defaultMode: 420 
  - name: goserver-vol 
    emptyDir: {} 
  - name: tmp-volume 
    emptyDir: {} 
  - name: tmp-conf-volume 
    emptyDir: {}

initContainers: 
- name: init-container 
image: artifactory-docker.artifactory.cib.echonet/gocd/gocd-server:v18.1.0 
command: - /bin/sh - '-c' - cp -r /go-server/* /mnt 
resources: 
limits: 
  cpu: '1' 
  memory: 3Gi requests: 
cpu: '1' memory: 2Gi 

volumeMounts: 
- name: tmp-conf-volume 
mountPath: /mnt 
terminationMessagePath: /dev/termination-log 
terminationMessagePolicy: File 
imagePullPolicy: IfNotPresent 
securityContext: 
readOnlyRootFilesystem: true 
allowPrivilegeEscalation: false 
containers: 
- name: gocd-server 
image: artifactory-docker.artifactory.cib.echonet/gocd/gocd-server:v18.1.0 
ports: - containerPort: 8153 
protocol: TCP 
resources: limits: cpu: '1' memory: 3Gi requests: cpu: '1' memory: 2Gi 
volumeMounts: 
- name: tmp-conf-volume
  mountPath: /go-server
```

let me explain the idea here of mounting volume. 
as i said the container dosent have the right to write on container disk, so whenever gocd tries to write or create a dir i need to mount external volume on that path  => I'm using Emptyvolume for the moment and i will replace them with real external volumes later.
Another tricky part In some a directory that have files from gocd image and generates file in the same directory. the idea here to get 
 the write right is to copy initial dir into a another volume and override it  (that's what i did in init container part).
let me know if you want more details? 

One last note: this version of gocd was deployed 3 years ago on rancher and it worked well, I'm trying to migrate it to kubernetes but as i said i don't have really a good vision about the issue here.

does it ring a beel on your side, am almost sure that the problem is related to volume mounting but what i can't understand the go.jar is there and why we are getting classNotFoundException, the logically is in the jar file

**Reply by hakim-dc** [reply]
another thing i'm not using pre configure option i want an empty gocd (i will paste the xml conf from the old instance)

**Reply by chadlwilson** [reply]
OK, so seems you are using a manual/custom Helm chart deployment rather than GoCD's own Helm chart. That is probably going to be a challenge. Getting volume mounts right for GoCD is important.

However a much bigger problem is I can't really help much with GoCD `18.x` - that is over 6 years old and there have been many major changes to the way the containers work since then. If you are running a GoCD that old you have **far bigger security concerns** than whether or not containers have read only filesystems :-)

`readOnlyRootFilesystem` is likely to be a blocker. The server does symlinking of working dirs at startup, and the agent downloads files to places that need layering on top of directories in the base image (e.g `/go`). If you are getting these errors, what is most likely happening is that you are mounting an `emptyDir` or other volume over the top of a location that GoCD expects things to sit, and now it cannot see them.

You should treat running with `readOnlyRootFilesystem` as unsupported, and while it irritates me personally, the container images have not been structured/designed for this.

**Reply by chadlwilson** [reply]
In other words, _personally_ I'm not able to go down a rabbithole of hacking around with `initContainers` and random mounts to workaround something GoCD images are not currently designed to do. I'd rather put the effort into making them support `readOnlyRootFilesystem` in the first place (not currently planned or something I was looking at though).

If you want to do this, you should probably build your own container image from either using gocd-server/agent base images, or using https://github.com/gocd/docker-gocd-server and https://github.com/gocd/docker-gocd-agent-wolfi (or similar) for inspiration ... or completely custom where you install the server/agent using an rpm/deb and then figure out your own process for getting the layout in the right state for a `readOnlyRootFilesystem` and a way to configure it.
